src/solver/conv_hip_implicit_gemm_fwd_v4r5_xdlops.cpp

```diff
@@ -0,0 +1,1042 @@
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/conv/invokers/impl_gemm.hpp>
+#include <miopen/solver.hpp>
+#include <miopen/handle.hpp>
+#include <miopen/generic_search.hpp>
+#include <miopen/hip_build_utils.hpp>
+#include "implicitgemm_util.hpp"
+/* this fix is for fp16 xdlops vectorizable kernels due to followings, we may revisit this fix after
+  compiler fix:
+  1. compiler issues(25% impact)
+  2. LDS write performance(75% impact)
+*/
+MIOPEN_DECLARE_ENV_VAR(
+    MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_FWD_V4R5_XDLOPS_ADD_VECTOR_LOAD_GEMMN_TUNE_PARAM)
+
+namespace miopen {
+namespace solver {
+
+PerformanceImplicitGemmForwardV4R5Xdlops::PerformanceImplicitGemmForwardV4R5Xdlops()
+    : PerformanceImplicitGemmForwardV4R5Xdlops::PerformanceImplicitGemmForwardV4R5Xdlops(
+          4, 4, 1, 4, 4, 1, false, false, 1)
+{
+}
+
+PerformanceImplicitGemmForwardV4R5Xdlops::PerformanceImplicitGemmForwardV4R5Xdlops(
+    int GemmMPerBlock_,
+    int GemmNPerBlock_,
+    int GemmKPerBlock_,
+    int GemmMPerWave_,
+    int GemmNPerWave_,
+    int GemmKPack_,
+    bool GemmAThreadCopyMoreGemmK_,
+    bool GemmBThreadCopyMoreGemmKPack_,
+    int GemmBThreadDataPerRead_GemmN_)
+    : GemmMPerBlock(GemmMPerBlock_),
+      GemmNPerBlock(GemmNPerBlock_),
+      GemmKPerBlock(GemmKPerBlock_),
+      GemmMPerWave(GemmMPerWave_),
+      GemmNPerWave(GemmNPerWave_),
+      GemmKPack(GemmKPack_),
+      GemmAThreadCopyMoreGemmK(GemmAThreadCopyMoreGemmK_),
+      GemmBThreadCopyMoreGemmKPack(GemmBThreadCopyMoreGemmKPack_),
+      GemmBThreadDataPerRead_GemmN(GemmBThreadDataPerRead_GemmN_)
+{
+}
+
+bool PerformanceImplicitGemmForwardV4R5Xdlops::
+operator==(const PerformanceImplicitGemmForwardV4R5Xdlops& other) const
+{
+    // clang-format off
+    return GemmMPerBlock == other.GemmMPerBlock
+        && GemmNPerBlock == other.GemmNPerBlock
+        && GemmKPerBlock == other.GemmKPerBlock
+        && GemmMPerWave == other.GemmMPerWave
+        && GemmNPerWave == other.GemmNPerWave
+        && GemmKPack == other.GemmKPack 
+        && GemmAThreadCopyMoreGemmK  == other.GemmAThreadCopyMoreGemmK
+        && GemmBThreadCopyMoreGemmKPack  == other.GemmBThreadCopyMoreGemmKPack
+        && GemmBThreadDataPerRead_GemmN  == other.GemmBThreadDataPerRead_GemmN;
+    // clang-format on
+}
+
+bool PerformanceImplicitGemmForwardV4R5Xdlops::SetNextValue()
+{
+    do
+    {
+        // list performance parameters in reverse order, in order for tuning to iterate over the
+        // range in normal order
+        if(miopen::IsEnabled(
+               MIOPEN_DEBUG_CONV_IMPLICIT_GEMM_FWD_V4R5_XDLOPS_ADD_VECTOR_LOAD_GEMMN_TUNE_PARAM{}))
+        {
+            if(!NextTwoPower<1, 8>(GemmBThreadDataPerRead_GemmN))
+                break;
+        }
+        if(!NextFlag<false, true>(GemmBThreadCopyMoreGemmKPack))
+            break;
+        if(!NextFlag<false, false>(GemmAThreadCopyMoreGemmK))
```

Comment:
@asroy Should <false, true>?

Reply:
This is based on experiment, you can try different values and fixate on one that gives best performance

Reply:
[Recommendation][Future]  Instead of the `if(!NextFlag<false, false>...)` stuff, the `GemmAThreadCopyMoreGemmK` should be initialized to `false`. That comment should tell the reader not what it is doing (that should be clear from the code) but WHY we want to do so. For example, in this specific case, we already know that fixation of parameter reduces the search space. What is interesting is:
- Why reducing of the search space is acceptable. I guess the reason is that parameter has very small effect.
- Why we want to keep the parameter instead of removing it. I guess we would like to make some experiments in the future.

Reply:
> * Why reducing of the search space is acceptable. I guess the reason is that parameter has very small effect.
> * Why we want to keep the parameter instead of removing it. I guess we would like to make some experiments in the future.

That's true. We didn't have time to looking to why some choice of (fixated) performance parameter always give better performance, but we don't want to lose such knowledge, so we keep them as tuning parameter but fixate the value. I agree we should add comment where it's initialized, and remove  NextFlag in the future.